// @mui/material/utils/index.js
export { default as createSvgIcon } from './createSvgIcon.js';
export { default as createChainedFunction } from '../../utils/esm/createChainedFunction.js';
export { default as debounce } from '../../utils/esm/debounce.js';
export { default as deprecatedPropType } from '../../utils/esm/deprecatedPropType.js';
export { default as isMuiElement } from '../../utils/esm/isMuiElement.js';
export { default as ownerDocument } from '../../utils/esm/ownerDocument.js';
export { default as ownerWindow } from '../../utils/esm/ownerWindow.js';
export { default as setRef } from '../../utils/esm/setRef.js';
export { default as unstable_useEnhancedEffect } from '../../utils/esm/useEnhancedEffect.js';
export { default as unstable_useId } from '../../utils/esm/useId.js';
export { default as unsupportedProp } from '../../utils/esm/unsupportedProp.js';
export { default as useControlled } from '../../utils/esm/useControlled.js';
export { default as useEventCallback } from '../../utils/esm/useEventCallback.js';
export { default as useForkRef } from '../../utils/esm/useForkRef.js';
export { default as useIsFocusVisible } from '../../utils/esm/useIsFocusVisible.js';
export { default as capitalize } from '../../utils/esm/capitalize.js';
export { default as requirePropFactory } from '../../utils/esm/requirePropFactory.js';
export { default as unstable_ClassNameGenerator } from '../../base/generateUtilityClass/ClassNameGenerator.js';
