// @mui/material/styles/createTheme.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('../../../@babel/runtime/helpers/esm/extends.js');
var objectWithoutPropertiesLoose = require('../../../@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js');
var system = require('@mui/system');
var createMixins = require('./createMixins.js');
var createPalette = require('./createPalette.js');
var createTypography = require('./createTypography.js');
var shadows = require('./shadows.js');
var createTransitions = require('./createTransitions.js');
var zIndex = require('./zIndex.js');
var deepmerge = require('../../utils/esm/deepmerge.js');

const _excluded = ["breakpoints", "mixins", "spacing", "palette", "transitions", "typography", "shape"];

function createTheme(options = {}, ...args) {
  const {
    mixins: mixinsInput = {},
    palette: paletteInput = {},
    transitions: transitionsInput = {},
    typography: typographyInput = {}
  } = options,
        other = objectWithoutPropertiesLoose["default"](options, _excluded);

  const palette = createPalette["default"](paletteInput);
  const systemTheme = system.createTheme(options);
  let muiTheme = deepmerge["default"](systemTheme, {
    mixins: createMixins["default"](systemTheme.breakpoints, systemTheme.spacing, mixinsInput),
    palette,
    // Don't use [...shadows] until you've verified its transpiled code is not invoking the iterator protocol.
    shadows: shadows["default"].slice(),
    typography: createTypography["default"](palette, typographyInput),
    transitions: createTransitions["default"](transitionsInput),
    zIndex: _extends["default"]({}, zIndex["default"])
  });
  muiTheme = deepmerge["default"](muiTheme, other);
  muiTheme = args.reduce((acc, argument) => deepmerge["default"](acc, argument), muiTheme);

  return muiTheme;
}

exports["default"] = createTheme;
